Solution for Issue 1977

tbowers7 opened the pull request:
The main issue in https://github.com/pypeit/PypeIt/issues/1977 is that sometimes (_e.g._, Soar/Goodman) raw files **MUST** be in some non-standard format (_e.g._, ".fz"), but the extension checker then throws out the processed files (_e.g._, "spec2d_*.fits") and crashes the post-processing scripts.

This solution adds "*.fits" to the ``spectrograph.allowed_extensions`` attribute **_within the confines of post-processing scripts_**.  So, when the ``Coadd2D`` script instantiates the appropriate ``Spectrograph`` class, ".fits" is added to the ``allowed_extensions`` attribute if not already present.

~~This may be a bit kludgy still, but removes the explicit reference to "spec2d*" files added in https://github.com/pypeit/PypeIt/pull/1976 .~~

The fix I settled on was  to add a "``pypeit_fits``" flag to the ``load_spectrograph()`` calls that sets the  ``allowed_extensions`` to ``[".fits"]``.  _If anyone has a better suggestion for the name of this flag, I'm all 👂._

tbowers7 committed:
Possible fix to Issue 1977

The main issue in https://github.com/pypeit/PypeIt/issues/1977 is that
sometimes (e.g., Soar/Goodman) raw files MUST be in some non-standard
format (e.g., ".fz"), but the extension checker then throws out the
processed files (e.g., "spec2d_*.fits") and crashes the post-processing
scritps.

This solution adds "*.fits" to the spectrograph.allowed_extensions
attribute *within the confines of post-processing scripts*.  So, when
the Coadd2D script instantiates the appropriate Spectrograph class,
".fits" is added to the allowed_extensions attribute if not already
present.

This may be a bit kludgy still, but removes the explicit reference
to "spec2d*" files added in https://github.com/pypeit/PypeIt/pull/1976 .

Another possible fix would be to add a "postproc" flag to the
load_spectrograph() calls, but that would require a little more fussing
with Coadd2D since it uses an input file -- all other places where I
added the ".fits" allowed_extensions were directly following a
load_spectrograph() call.  Something to discuss / think about.

	modified:   pypeit/scripts/coadd_1dspec.py
	modified:   pypeit/scripts/coadd_2dspec.py
	modified:   pypeit/scripts/coadd_datacube.py
	modified:   pypeit/scripts/collate_1d.py
	modified:   pypeit/scripts/flux_calib.py
	modified:   pypeit/scripts/flux_setup.py
	modified:   pypeit/scripts/multislit_flexure.py
	modified:   pypeit/scripts/sensfunc.py
	modified:   pypeit/scripts/tellfit.py
	modified:   pypeit/spectrographs/spectrograph.py

tbowers7 committed:
Stay focused...

Don't try to clean everyone else's code.  Just stick to the point.

	modified:   pypeit/scripts/coadd_2dspec.py
	modified:   pypeit/spectrographs/spectrograph.py

rcooke-ast reviewed (approved):
I think this is a great solution - much simpler 👍 Happy to approve this, but I raise an alternative to consider as well:

We could replace these if statements with:
`spectrograph.check_addfits()`

and then add the following method to the spectrograph class:

```
def check_addfits(self):
    """
    Check if the fits extension should be added to the allowed extensions
    """
    if '.fits' not in self.allowed_extensions:
        self.allowed_extensions.append('.fits')
```

The only benefit with this approach would be if we change the default (output) extensions to be something like `.fits.gz` or some equivalent. In this case, we would only need to make one change (to the spectrograph class). The downside, is that we wouldn't want this definition to be called from basically any other part of the code, and we wouldn't want child classes to overwrite. On balance, I'd be happy with either!

tbowers7 committed:
Move the extension change to ``load_spectrograph()``

Instead of having each script need to have code adding the ".fits"
extension for post-processing, just do it all in ``load_spectrograph()``
and activate that feature with a boolean "pypeit_fits" flag.

	modified:   doc/releases/1.18.2dev.rst
	modified:   pypeit/scripts/coadd_1dspec.py
	modified:   pypeit/scripts/coadd_2dspec.py
	modified:   pypeit/scripts/coadd_datacube.py
	modified:   pypeit/scripts/collate_1d.py
	modified:   pypeit/scripts/flux_calib.py
	modified:   pypeit/scripts/flux_setup.py
	modified:   pypeit/scripts/multislit_flexure.py
	modified:   pypeit/scripts/sensfunc.py
	modified:   pypeit/scripts/tellfit.py
	modified:   pypeit/spectrographs/util.py

tbowers7 commented:
@rcooke-ast Thanks for the review.  After the initial version failed some tests, I decided to go with the flag in ``load_spectrograph()`` route.  I'd appreciate a look-see at the new edits.

```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS FAILED  1 failed, 274 passed, 471 warnings in 477.04s (0:07:57) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 93527 warnings in 1036.15s (0:17:16) ---
--- PYTEST VET TESTS FAILED  3 failed, 66 passed, 153344 warnings in 6912.30s (1:55:12) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 2/281 TESTS  ---
Failed tests:
    gemini_gmos/GS_HAM_R400_700 pypeit_tellfit
    gemini_gnirs_echelle/32_SB_SXD pypeit_tellfit
Skipped tests:
Testing Started at 2025-10-11T01:17:47.437392
Testing Completed at 2025-10-11T20:12:40.392579
Total Time: 18:54:52.955187
```
New DevSuite run in progress...

rcooke-ast commented on pypeit/spectrographs/util.py:
```diff
@@ -25,6 +25,11 @@ def load_spectrograph(spec):
+        pypeit_fits (:obj:`bool`, optional):
+            The spectrograph loader is being called from a post-processing
+            script where the expected input files are PypeIt-written FITS files
+            only.  This has the effect of overriding the :attr:`allowed_extensions``
```
missing one left quote before allowed_extensions

tbowers7 committed:
Fix testing errors; add unit test

Fix the ``load_spectrograph()`` location that I missed for updating
the allowed extensions.  Coadd1d files do not incude the [rdx] section,
so can't reference that.

Added unit test for the ``load_spectrograph()`` function that tests all
three methods for instantiating a Spectrograph class and also tests the
new "pypeit_fits" boolean flag at the heart of this PR (#2013).

	modified:   pypeit/scripts/coadd_1dspec.py
	modified:   pypeit/spectrographs/util.py
	modified:   pypeit/tests/test_spectrographs.py

tbowers7 commented:
Running the DevSuite one more time...

tbowers7 committed:
Fix test failures

Will probably add an Issue about Gemini/GNIRS using ``self.dispname``
for distinguishing between their gratings -- it is the only spectrograph
to do so, and there is a bit of annoying code splattered about through
PypeIt to deal with it.

Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  276 passed, 471 warnings in 495.56s (0:08:15) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 93527 warnings in 1077.75s (0:17:57) ---
--- PYTEST VET TESTS FAILED  4 failed, 65 passed, 138629 warnings in 7110.57s (1:58:30) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 2/281 TESTS  ---
Failed tests:
    soar_goodman_blue/M1 pypeit_coadd_2dspec
    gemini_gnirs_echelle/32_SB_SXD pypeit_coadd_2dspec
Skipped tests:
    gemini_gnirs_echelle/32_SB_SXD pypeit_tellfit
Testing Started at 2025-10-13T06:11:07.662789
Testing Completed at 2025-10-14T01:35:14.834860
Total Time: 19:24:07.172071

	modified:   pypeit/inputfiles.py
	modified:   pypeit/scripts/coadd_2dspec.py
	modified:   pypeit/scripts/ql.py

tbowers7 committed:
Fix coadd2d output filenames to remove spaces

This is standard for regular output files; make it the same for
coadd2d.

	modified:   pypeit/coadd2d.py

tbowers7 commented:
DevSuite tests pass (`tbowers-iss1977-14oct.report`).  Ready for review.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  276 passed, 471 warnings in 503.75s (0:08:23) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 93527 warnings in 1274.79s (0:21:14) ---
--- PYTEST VET TESTS PASSED  69 passed, 145856 warnings in 6753.40s (1:52:33) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 282/282 TESTS  ---
Testing Started at 2025-10-14T17:19:39.202315
Testing Completed at 2025-10-15T10:08:49.012127
Total Time: 16:49:09.809812
```

rcooke-ast commented:
Still looks good to me @tbowers7 - thanks!

tbowers7 committed:
Merge branch 'develop' into issue_1977

tbowers7 committed:
Merge branch 'develop' into issue_1977

tbowers7 committed:
Merge branch 'develop' into issue_1977

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -318,7 +318,7 @@ def default_basename(spec2d_files):
                         'using the command-line option.')
         return f"{frsthdr['FILENAME'].split('.fits')[0]}-" \
-                f"{lasthdr['FILENAME'].split('.fits')[0]}-{frsthdr['TARGET']}"
+                f"{lasthdr['FILENAME'].split('.fits')[0]}-{frsthdr['TARGET'].replace(' ','')}"
```
Nice.  Good edit!

kbwestfall commented on pypeit/tests/test_spectrographs.py:
```diff
@@ -148,4 +148,39 @@ def test_atmext():
+
+    # Call using a single processed data file, and from a post-processing script
+    spec8 = load_spectrograph(raw_file, pypeit_fits=True)
+    assert spec8.allowed_extensions == [".fits"], 'Postproc scripts only allow .fits'
```
Thanks for the new tests!

kbwestfall reviewed (approved):
I'm fine with this, but we should still keep this issue in mind for future refactors. Thanks!

tbowers7 committed:
Merge branch 'develop' into issue_1977

tbowers7 committed:
Merge branch 'develop' into issue_1977

tbowers7 committed:
Merge branch 'develop' into issue_1977

kbwestfall committed:
Merge branch 'develop' into issue_1977
